content-services/content-sources-backend
# Fixes 4741: associate snapshots to templates directly

## Summary

- Associates snapshots to templates by adding snapshot_uuid to the templates_repository_configurations table
- Migrates any existing templates' snapshots to this table
- Simplifies the query for listing a template's snapshots to use the template-snapshot association
- Checks that repos added to a template have a snapshot and rejects request if they do not

## Testing steps

To test the migration:

1. In main, create 1 custom repo and 1 RH repo
2. Change the URL of the custom repo so you have 2 snapshots, and update the snapshot's created_at dates to be on different days
3. Change the URL one more time to get another snapshot and don't update this date
4. Create 3 templates: 1 with a date set to be between the dates of the snapshots, 1 with a date before the snapshots, and 1 using latest content
5. Check out this PR and run the migration. Verify the correct snapshot uuids were added to the templates_repository_configurations table (you can either check the table directly or check via the `GET /templates/:uuid/snapshots/` endpoint)

To test a template's snapshots are updated correctly:

1. Create a custom repo and a RH repo, let them snapshot, and create a template with these repos with use_latest set to true
2. Listing the template's snapshots (`GET /templates/:uuid/snapshots/`) should still work as expected
3. Add / remove one of the template's repos, listing the template's snapshots show updated snapshots
4. Trigger another snapshot of one of the repos in template. Once this task completes you should see that latest snapshot when listing the template's snapshots
